pkg/ptrace/ptrace.go

```diff
@@ -0,0 +1,484 @@
+	}
+
+	traceSuccess = true
+
+	return program, nil
+}
+
+// Detach detaches from all threads of the processs
+func (p *TracedProgram) Detach() error {
+	for _, tid := range p.tids {
+		err := syscall.PtraceDetach(tid)
+
+		if err != nil {
+			log.Error(err, "detach failed", "tid", tid)
+			return errors.WithStack(err)
+		}
+	}
+
+	log.Info("Successfully detach and rerun process", "pid", p.pid)
+	return nil
+}
+
+// Protect will backup regs and rip into fields
+func (p *TracedProgram) Protect() error {
+	err := syscall.PtraceGetRegs(p.pid, p.backupRegs)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	_, err = syscall.PtracePeekData(p.pid, uintptr(p.backupRegs.Rip), p.backupCode)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+// Restore will restore regs and rip from fields
+func (p *TracedProgram) Restore() error {
+	err := syscall.PtraceSetRegs(p.pid, p.backupRegs)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	_, err = syscall.PtracePokeData(p.pid, uintptr(p.backupRegs.Rip), p.backupCode)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+// Wait waits until the process stops
+func (p *TracedProgram) Wait() error {
+	return waitPid(p.pid)
+}
+
+// Step moves one step forward
+func (p *TracedProgram) Step() error {
+	err := syscall.PtraceSingleStep(p.pid)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return p.Wait()
+}
+
+// Syscall runs a syscall at main thread of process
+func (p *TracedProgram) Syscall(number uint64, args ...uint64) (uint64, error) {
+	err := p.Protect()
+	if err != nil {
+		return 0, err
+	}
+
+	var regs syscall.PtraceRegs
+
+	err = syscall.PtraceGetRegs(p.pid, &regs)
+	if err != nil {
+		return 0, err
+	}
+	regs.Rax = number
+	for index, arg := range args {
+		// All these registers are hard coded for x86 platform
+		if index == 0 {
+			regs.Rdi = arg
+		} else if index == 1 {
+			regs.Rsi = arg
+		} else if index == 2 {
+			regs.Rdx = arg
+		} else if index == 3 {
+			regs.R10 = arg
+		} else if index == 4 {
+			regs.R8 = arg
+		} else if index == 5 {
+			regs.R9 = arg
+		} else {
+			return 0, fmt.Errorf("too many arguments for a syscall")
+		}
+	}
+	err = syscall.PtraceSetRegs(p.pid, &regs)
+	if err != nil {
+		return 0, err
+	}
+
+	ip := make([]byte, ptrSize)
+
+	// We only support x86-64 platform now, so using hard coded `LittleEndian` here is ok.
+	binary.LittleEndian.PutUint16(ip, 0x050f)
+	_, err = syscall.PtracePokeData(p.pid, uintptr(p.backupRegs.Rip), ip)
+	if err != nil {
+		return 0, err
+	}
+
+	err = p.Step()
+	if err != nil {
+		return 0, err
+	}
+
+	err = syscall.PtraceGetRegs(p.pid, &regs)
+	if err != nil {
+		return 0, err
+	}
+
+	return regs.Rax, p.Restore()
+}
+
+// Mmap runs mmap syscall
+func (p *TracedProgram) Mmap(length uint64, fd uint64) (uint64, error) {
+	return p.Syscall(9, 0, length, syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC, syscall.MAP_ANON|syscall.MAP_PRIVATE, fd, 0)
```

Comment:
why is the number `9`?

Reply:
It is replaced with `syscall.SYS_MMAP` now :beers: